### PR TITLE
DP-1015 customizable search in cell provider

### PR DESCRIPTION
**Pull request by CheatEx**

### CODE REVIEW

**Comment by CheatEx** [comment]
@katepol @timzam After discussion with Gleb we concluded that there should be a direct `CellProvider` test.

Also I hit an overloading problem: call `getCells(List)` gives unexpected result due to interpretation of list as a path. This commit also includes a possible renaming solutions for that problem, I propose to delete `@Deprecated` methods.

File: projectional/src/test/java/jetbrains/jetpad/projectional/cell/mapping/CellProviderTest.java [hunk 1]
```diff
@@ -0,0 +1,115 @@
+package jetbrains.jetpad.projectional.cell.mapping;
```
Comment (timzam) [inline projectional/src/test/java/jetbrains/jetpad/projectional/cell/mapping/CellProviderTest.java]: Please add the license.

**Comment by timzam** [comment]
Looks good to me.

**Comment by CheatEx** [comment]
@timzam done!

File: projectional/src/main/java/jetbrains/jetpad/projectional/cell/mapping/CellProvider.java [hunk 2]
```diff
@@ -70,30 +72,51 @@ public void remove() {
...
+   *   search continues up through cell hierarchy.
+   */
+  public Object getSource(Cell cell, Predicate<Object> predicate) {
     for (Cell c = cell; c != null; c = c.getParent()) {
       Collection<Mapper<?, ?>> mappers = myIndex.getMappers(c);
       if (!mappers.isEmpty()) {
         Mapper<?, ?> mapper = mappers.iterator().next();
         Object source = doGetSource(mapper, cell);
-        return source == null ? mapper.getSource() : source;
+        Object res = source == null ? mapper.getSource() : source;
```
Comment (katepol) [inline projectional/src/main/java/jetbrains/jetpad/projectional/cell/mapping/CellProvider.java]: If by 'res' you mean 'result', why don't you just put this name directly?

File: projectional/src/main/java/jetbrains/jetpad/projectional/cell/mapping/CellProvider.java [hunk 3]
```diff
@@ -148,7 +171,7 @@ private Object getSourceFromProjectionalSynchronizer(ProjectionalRoleSynchronize
 
   private List<Cell> doGetCells(Mapper<?, ? extends Cell> mapper, Object actualSource) {
     if (mapper.getSource() == actualSource) {
-      return Collections.<Cell>singletonList(mapper.getTarget());
+      return Collections.singletonList(mapper.getTarget());
```
Comment (katepol) [inline projectional/src/main/java/jetbrains/jetpad/projectional/cell/mapping/CellProvider.java]: GWT needs an explicit type parameter here for successful compilation. Get it back, please.

**Comment by azlobinjb** [comment]
@katepol done.

**Pull request merged**